File: go/generate.go
// SPDX-License-Identifier: CC-BY-4.0
// Copyright (c) 2025-2026 fumi-engineer

package nn

import "sort"

// SamplingStrategy defines how to pick the next token from logit scores.
// Implementations: GreedySampling, TemperatureSampling, TopKSampling, TopPSampling.
type SamplingStrategy interface {
	PickToken(logits []float32) int
}

// GreedySampling always picks the token with the highest logit (argmax).
type GreedySampling struct{}

// PickToken returns the index of the maximum logit.
func (g GreedySampling) PickToken(logits []float32) int {
	idx, _ := argmax(logits)
	return idx
}

// TemperatureSampling scales logits by 1/T then samples from the softmax
// distribution. Higher T = more random, lower T = more greedy.
type TemperatureSampling struct {
	Temperature float32
	State       *uint64 // PRNG state (LCG)
}

// PickToken samples a token after temperature scaling.
func (s TemperatureSampling) PickToken(logits []float32) int {
	return sampleFromLogits(logits, s.Temperature, s.State)
}

// TopKSampling restricts sampling to the K highest-probability tokens,
// then applies temperature sampling within that subset.
type TopKSampling struct {
	K           int
	Temperature float32
	State       *uint64
}

// PickToken samples from the top-K logits.
func (s TopKSampling) PickToken(logits []float32) int {
	return sampleTopKFromLogits(logits, s.K, s.Temperature, s.State)
}

// TopPSampling (nucleus sampling) includes the smallest set of tokens whose
// cumulative probability exceeds TopP, then samples from that subset.
type TopPSampling struct {
	TopP        float32
	Temperature float32
	State       *uint64
}

// PickToken samples from the nucleus (top-p) of the distribution.
func (s TopPSampling) PickToken(logits []float32) int {
	return sampleTopPFromLogits(logits, s.TopP, s.Temperature, s.State)
}

// Generate produces tokens auto-regressively using the given sampling strategy.
// At each step, the full sequence is fed through the model (no KV cache),
// the last position's logits are sampled, and the new token is appended.
func Generate(m *MoETransformer, prompt []int, maxLen int, strategy SamplingStrategy) []int {
	tokens := cloneInts(prompt)
	if len(tokens) == 0 {
		tokens = append(tokens, 0)
	}
	for len(tokens) < maxLen {
		logits := m.ForwardIDs(tokens, 1, len(tokens))
		// Extract logits for the last position only.
		// Flat offset: (seq_len - 1) * vocab_size
		last := logits.DataPtr()[(len(tokens)-1)*m.config.VocabSize : len(tokens)*m.config.VocabSize]
		tokens = append(tokens, strategy.PickToken(last))
	}
	return tokens
}

// GenerateGreedy is a convenience method for greedy (argmax) decoding.
func (m *MoETransformer) GenerateGreedy(prompt []int, maxLen int) []int {
	return Generate(m, prompt, maxLen, GreedySampling{})
}

// GenerateSample generates with temperature sampling.
func (m *MoETransformer) GenerateSample(prompt []int, maxLen int, temperature float32, seed uint64) []int {
	return Generate(m, prompt, maxLen, TemperatureSampling{Temperature: temperature, State: &seed})
}

// GenerateTopK generates with top-K sampling.
func (m *MoETransformer) GenerateTopK(prompt []int, maxLen, k int, temperature float32, seed uint64) []int {
	return Generate(m, prompt, maxLen, TopKSampling{K: k, Temperature: temperature, State: &seed})
}

// GenerateTopP generates with nucleus (top-P) sampling.
func (m *MoETransformer) GenerateTopP(prompt []int, maxLen int, topP, temperature float32, seed uint64) []int {
	return Generate(m, prompt, maxLen, TopPSampling{TopP: topP, Temperature: temperature, State: &seed})
}

// softmaxInPlace applies softmax to xs in-place with numerical stability.
//   p_i = exp(x_i - max(x)) / sum(exp(x_j - max(x)))
func softmaxInPlace(xs []float32) {
	if len(xs) == 0 {
		return
	}
	_, maxVal := argmax(xs)
	for i := range xs {
		xs[i] = ExpF32(xs[i] - maxVal)
	}
	normalizeInPlace(xs)
}

// nextRand01 returns a pseudo-random float32 in [0, 1) using a 64-bit LCG.
// The multiplier is from Knuth's MMIX. Using a simple LCG avoids importing
// math/rand, keeping sampling deterministic and reproducible from a seed.
func nextRand01(state *uint64) float32 {
	*state = *state*6364136223846793005 + 1
	return float32(uint32(*state>>32)) / 4294967296.0
}

// sampleFromProbs samples an index from a discrete probability distribution
// using the inverse CDF method.
func sampleFromProbs(probs []float32, state *uint64) int {
	r := nextRand01(state)
	cum := float32(0)
	for i, p := range probs {
		cum += p
		if r <= cum {
			return i
		}
	}
	return len(probs) - 1
}

// sampleFromLogits applies temperature scaling, softmax, then samples.
//   scaled_i = logit_i / temperature
//   p = softmax(scaled)
//   sample from p
func sampleFromLogits(logits []float32, temperature float32, state *uint64) int {
	if temperature <= 0 {
		idx, _ := argmax(logits)
		return idx
	}
	scaled := make([]float32, len(logits))
	invTemp := float32(1.0) / temperature
	for i, v := range logits {
		scaled[i] = v * invTemp
	}
	softmaxInPlace(scaled)
	return sampleFromProbs(scaled, state)
}

// sampleTopKFromLogits keeps only the K highest logits (setting the rest to -inf),
// then delegates to temperature sampling on the filtered distribution.
func sampleTopKFromLogits(logits []float32, k int, temperature float32, state *uint64) int {
	n := len(logits)
	if k <= 0 || k >= n {
		return sampleFromLogits(logits, temperature, state)
	}

	indices := make([]int, n)
	for i := range indices {
		indices[i] = i
	}
	sort.Slice(indices, func(i, j int) bool {
		return logits[indices[i]] > logits[indices[j]]
	})

	filtered := make([]float32, n)
	for i := range filtered {
		filtered[i] = NegInf
	}
	for i := 0; i < k; i++ {
		filtered[indices[i]] = logits[indices[i]]
	}
	return sampleFromLogits(filtered, temperature, state)
}

// sampleTopPFromLogits implements nucleus sampling: accumulates tokens from
// highest to lowest probability until the cumulative probability exceeds topP,
// then samples from that nucleus.
func sampleTopPFromLogits(logits []float32, topP, temperature float32, state *uint64) int {
	if topP <= 0 || topP >= 1 {
		return sampleFromLogits(logits, temperature, state)
	}
	if temperature <= 0 {
		idx, _ := argmax(logits)
		return idx
	}

	scaled := make([]float32, len(logits))
	invTemp := float32(1.0) / temperature
	for i, v := range logits {
		scaled[i] = v * invTemp
	}
	softmaxInPlace(scaled)

	indices := make([]int, len(logits))
	for i := range indices {
		indices[i] = i
	}
	sort.Slice(indices, func(i, j int) bool {
		return scaled[indices[i]] > scaled[indices[j]]
	})

	// Accumulate highest-probability tokens until sum >= topP
	selected := make([]int, 0, len(indices))
	sum := float32(0)
	for _, idx := range indices {
		selected = append(selected, idx)
		sum += scaled[idx]
		if sum >= topP {
			break
		}
	}

	trunc := make([]float32, len(selected))
	for i, idx := range selected {
		trunc[i] = scaled[idx]
	}
	normalizeInPlace(trunc)

	chosen := sampleFromProbs(trunc, state)
	return selected[chosen]
}

File: python/generate.py
# SPDX-License-Identifier: CC-BY-NC-4.0
# Copyright (c) 2025-2026 fumi-engineer

"""Sampling strategies and token generation for MoE Transformer.

Provides four sampling strategies:
  - Greedy (argmax)
  - Temperature scaling
  - Top-k filtering
  - Top-p (nucleus) filtering

The generate() function runs autoregressive decoding: at each step it
feeds the full token sequence through the model (no KV cache), takes
the logits at the last position, samples a token, and appends it.

All random sampling uses a deterministic LCG PRNG for reproducibility
across platforms (independent of numpy's RNG state).
"""

from __future__ import annotations

from abc import ABC, abstractmethod

import numpy as np


class SamplingStrategy(ABC):
    """Abstract base class for sampling strategies."""

    @abstractmethod
    def pick_token(self, logits: np.ndarray) -> int:
        """Pick the next token given logits [vocab_size]."""
        ...


class GreedySampling(SamplingStrategy):
    """Greedy (argmax) sampling — always picks the highest-probability token."""

    def pick_token(self, logits: np.ndarray) -> int:
        return int(np.argmax(logits))


class TemperatureSampling(SamplingStrategy):
    """Temperature-based sampling.

    Sampling: p_i = softmax(logits_i / temperature)
    Higher temperature -> more uniform distribution (more random).
    Lower temperature -> sharper distribution (more deterministic).
    """

    def __init__(self, temperature: float = 1.0, seed: int = 42):
        self.temperature = temperature
        self._state = [seed]

    def pick_token(self, logits: np.ndarray) -> int:
        return _sample_from_logits(logits, self.temperature, self._state)


class TopKSampling(SamplingStrategy):
    """Top-k sampling.

    Filters to the k highest-probability tokens before sampling.
    All other tokens are masked to -inf so they receive zero probability.
    """

    def __init__(self, k: int = 50, temperature: float = 1.0, seed: int = 42):
        self.k = k
        self.temperature = temperature
        self._state = [seed]

    def pick_token(self, logits: np.ndarray) -> int:
        return _sample_top_k(logits, self.k, self.temperature, self._state)


class TopPSampling(SamplingStrategy):
    """Top-p (nucleus) sampling.

    Sorts tokens by probability and includes just enough tokens for
    their cumulative probability to exceed top_p.  This dynamically
    adjusts the candidate set size based on the distribution shape.
    """

    def __init__(self, top_p: float = 0.9, temperature: float = 1.0, seed: int = 42):
        self.top_p = top_p
        self.temperature = temperature
        self._state = [seed]

    def pick_token(self, logits: np.ndarray) -> int:
        return _sample_top_p(logits, self.top_p, self.temperature, self._state)


def generate(model, prompt: list[int], max_len: int, strategy: SamplingStrategy) -> list[int]:
    """Generate tokens autoregressively using a given sampling strategy.

    At each step, the full sequence is re-processed through the model
    (no KV cache), which is O(n^2) in sequence length.  This is
    intentionally simple for educational purposes.
    """
    tokens = list(prompt) if prompt else [0]
    while len(tokens) < max_len:
        logits = model.forward_ids(tokens, 1, len(tokens))
        # Take logits at the last position for next-token prediction
        last = logits.data[0, len(tokens) - 1]
        tokens.append(strategy.pick_token(last))
    return tokens


# Backward-compatible function wrappers

def generate_greedy(model, prompt: list[int], max_len: int) -> list[int]:
    """Generate tokens using greedy decoding."""
    return generate(model, prompt, max_len, GreedySampling())


def generate_sample(
    model, prompt: list[int], max_len: int, temperature: float = 1.0, seed: int = 42
) -> list[int]:
    """Generate tokens using temperature sampling."""
    return generate(model, prompt, max_len, TemperatureSampling(temperature, seed))


def generate_top_k(
    model,
    prompt: list[int],
    max_len: int,
    k: int = 50,
    temperature: float = 1.0,
    seed: int = 42,
) -> list[int]:
    """Generate tokens using top-k sampling."""
    return generate(model, prompt, max_len, TopKSampling(k, temperature, seed))


def generate_top_p(
    model,
    prompt: list[int],
    max_len: int,
    top_p: float = 0.9,
    temperature: float = 1.0,
    seed: int = 42,
) -> list[int]:
    """Generate tokens using top-p (nucleus) sampling."""
    return generate(model, prompt, max_len, TopPSampling(top_p, temperature, seed))


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------

def _next_rand01(state: list[int]) -> float:
    """LCG PRNG returning [0, 1).

    Uses a Linear Congruential Generator with the same constants as
    PCG's default multiplier.  State is a 1-element list for mutability
    (Python ints are immutable, so we mutate the list in-place).
    This provides cross-platform deterministic randomness independent
    of numpy's global RNG state.
    """
    state[0] = (state[0] * 6_364_136_223_846_793_005 + 1) & 0xFFFFFFFFFFFFFFFF
    return float((state[0] >> 32) & 0xFFFFFFFF) / 4294967296.0


def _softmax(xs: np.ndarray) -> np.ndarray:
    """Numerically stable softmax for 1D array.

    Softmax: p_i = exp(x_i - max(x)) / sum_j(exp(x_j - max(x)))
    """
    shifted = xs - np.max(xs)
    exp_vals = np.exp(shifted)
    total = np.sum(exp_vals)
    if total == 0:
        return exp_vals
    return exp_vals / total


def _sample_from_probs(probs: np.ndarray, state: list[int]) -> int:
    """Sample an index from a probability distribution via inverse CDF.

    Draws a uniform random number and walks the cumulative distribution
    until the threshold is crossed.
    """
    r = _next_rand01(state)
    cum = 0.0
    for i, p in enumerate(probs):
        cum += p
        if r <= cum:
            return i
    return len(probs) - 1


def _sample_from_logits(
    logits: np.ndarray, temperature: float, state: list[int]
) -> int:
    """Sample from logits with temperature scaling.

    scaled_logits = logits / temperature
    probs = softmax(scaled_logits)
    token = sample(probs)

    Temperature=0 degenerates to greedy (argmax).
    """
    if temperature <= 0:
        return int(np.argmax(logits))
    scaled = logits / temperature
    probs = _softmax(scaled)
    return _sample_from_probs(probs, state)


def _sample_top_k(
    logits: np.ndarray, k: int, temperature: float, state: list[int]
) -> int:
    """Sample from top-k filtered logits.

    Masks all tokens outside the top-k to -inf before applying
    temperature scaling and softmax.
    """
    n = len(logits)
    if k <= 0 or k >= n:
        return _sample_from_logits(logits, temperature, state)

    indices = np.argsort(logits)[::-1]
    filtered = np.full(n, -np.inf)
    for i in range(k):
        filtered[indices[i]] = logits[indices[i]]
    return _sample_from_logits(filtered, temperature, state)


def _sample_top_p(
    logits: np.ndarray, top_p: float, temperature: float, state: list[int]
) -> int:
    """Sample from top-p (nucleus) filtered logits.

    Sorts tokens by descending probability, includes tokens until the
    cumulative probability exceeds top_p, then renormalizes and samples.
    """
    if top_p <= 0 or top_p >= 1:
        return _sample_from_logits(logits, temperature, state)
    if temperature <= 0:
        return int(np.argmax(logits))

    scaled = logits / temperature
    probs = _softmax(scaled)

    # Sort by descending probability and accumulate until threshold
    indices = np.argsort(probs)[::-1]
    cum = 0.0
    selected = []
    for idx in indices:
        selected.append(idx)
        cum += probs[idx]
        if cum >= top_p:
            break

    # Renormalize the selected subset
    trunc = np.array([probs[idx] for idx in selected])
    total = np.sum(trunc)
    if total > 0:
        trunc /= total

    chosen = _sample_from_probs(trunc, state)
    return int(selected[chosen])
